src/lib/info_tree.rs

```diff
@@ -0,0 +1,431 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#![warn(missing_docs)]
+#![allow(dead_code)]  // Temporary, this module will be used in RTED.
+
+/// Stores information about a single AST (as needed by RTED) in vectors.
+
+use ast::{Arena, NodeId};
+use label_maps::LabelMap;
+
+const LEFT: u8 = 0;
+const RIGHT: u8 = 1;
+const HEAVY: u8 = 2;
+const BOTH: u8 = 3;
+
+const POST2_SIZE: u8 = 0;
+const POST2_KR_SUM: u8 = 1;
+const POST2_REV_KR_SUM: u8 = 2;
+// Number of subforests in full decomposition.
+const POST2_DESC_SUM: u8 = 3;
+const POST2_PRE: u8 = 4;
+const POST2_PARENT: u8 = 5;
+const POST2_LABEL: u8 = 6;
+// Key root nodes (size of this array = leaf count).
+const KR: u8 = 7;
+// Left-most leaf descendants.
+const POST2_LLD: u8 = 8;
+// Minimum key root nodes index in key root array.
+const POST2_MIN_KR: u8 = 9;
+// Reversed key root nodes.
+const RKR: u8 = 10;
+// Reversed postorer 2 right-most leaf descendants.
+const RPOST2_RLD: u8 = 11;
+const RPOST2_MIN_RKR: u8 = 12;
+// Reversed postorder -> postorder.
+const RPOST2_POST: u8 = 13;
+// Strategy for Demaine.
+const POST2_STRATEGY: u8 = 14;
+const PRE2_POST: u8 = 15;
+
+
+struct InfoTree<'a> {
+    // All information related to a given AST and necessary for RTED.
+    pub info: Vec<Vec<Option<usize>>>,
+    // Node labels stored as numbers (as an optimisation).
+    labels: LabelMap<'a>,
+    // Flags (left, right, heavy) for each node.
+    pub node_type: Vec<Vec<bool>>,
+    // Paths is a vector to be able to get them by paths/relsubtrees[L/R/H][node]
+    paths: Vec<Vec<Option<usize>>>,
+    // Rel subtrees is a vector to be able to get them by paths/relsubtrees[L/R/H][node]
+    relative_subtrees: Vec<Vec<Vec<usize>>>,
+    tmp_size: usize,
+    tmp_descendant_sizes: usize,
+    tmp_key_root_sizes: usize,
+    tmp_rev_key_root_sizes: usize,
+    tmp_preorder: usize,
+    // Postorder number of the current node in the TED recursion.
+    current_node: usize,
+    // Remember if the tree order was switched during the recursion (by
+    // comparison with the order of the input tree).
+    switched: bool,
+    leaf_count: usize,
+    tree_size: usize,
+}
+
+impl<'a> InfoTree<'a> {
+    /// Create a new information tree for a non-empty AST.
+    pub fn new<T: Clone + PartialEq, U: Copy + PartialEq>(ast: &'a Arena<T, U>, label_map: LabelMap<'a>) -> InfoTree<'a> {
+        assert!(ast.root().is_some(), "Cannot create an InfoTree on an empty Arena.");
+        let mut tree = InfoTree {
+            info: vec![],
+            labels: label_map,
+            node_type: vec![],
+            paths: vec![],
+            relative_subtrees: vec![],
+            tmp_size: 0,
+            tmp_descendant_sizes: 0,
+            tmp_key_root_sizes: 0,
+            tmp_rev_key_root_sizes: 0,
+            tmp_preorder: 0,
+            current_node: ast.size() - 1,
```

Comment:
I suspect that either we should assert above that `ast.size() > 0` or use `checked_sub`.

Reply:
On line 105 we have asserted that the AST has a root node. I'll add a comment here.